src/adapters/postproc.rs

```diff
@@ -145,38 +143,57 @@ pub fn postproc_prefix(line_prefix: &str, inp: impl AsyncRead + Send) -> impl As
     StreamReader::new(oup_stream)
 }
 
-/**
- * adds the prefix `Page N:` to each line,
- * where N starts at one and is incremented for each ASCII Form Feed character in the input stream.
- * (That's the format output by pdftotext)
- */
-pub fn postproc_pagebreaks(line_prefix: &str, inp: impl AsyncRead) -> impl AsyncRead {
-    let line_prefix = line_prefix.to_string(); // clone since
-    let mut page_count = 1;
+/// Adds the prefix "Page N:" to each line,
+/// where N starts at one and is incremented for each ASCII Form Feed character in the input stream.
+/// ASCII form feeds are the page delimiters output by `pdftotext`.
+pub fn postproc_pagebreaks(input: impl AsyncRead + Send) -> impl AsyncRead + Send {
+    let regex_linefeed = regex::bytes::Regex::new(r"\x0c").unwrap();
+    let regex_newline = regex::bytes::Regex::new("\n").unwrap();
+    let mut page_count: i32 = 1;
+    let mut page_prefix: String = format!("\nPage {}:", page_count);
 
-    panic!("todo!");
-    tokio::io::empty()
-    /*ByteReplacer {
-        inner: inp,
-        next_read: format!("{}Page {}:", line_prefix, page_count).into_bytes(),
-        haystacker: Box::new(|buf| memchr::memchr2(b'\n', b'\x0c', buf)),
-        replacer: Box::new(move |b| match b {
-            b'\n' => format!("\n{}Page {}:", line_prefix, page_count).into_bytes(),
-            b'\x0c' => {
-                page_count += 1;
-                format!("\n{}Page {}:", line_prefix, page_count).into_bytes()
+    let input_stream = ReaderStream::new(input);
+    let output_stream = stream! {
+        for await chunk in input_stream {
+            match chunk {
+                Err(e) => yield Err(e),
+                Ok(chunk) => {
+                    let sub_chunks = regex_linefeed.split(&chunk);
+                    for sub_chunk in sub_chunks {
+                        // println!("{}", String::from_utf8_lossy(page_prefix.as_bytes()));
+                        yield Ok(Bytes::copy_from_slice(page_prefix.as_bytes()));
+                        yield Ok(Bytes::copy_from_slice(&regex_newline.replace_all(&sub_chunk, page_prefix.as_bytes())));
+                        page_count += 1;
+                        page_prefix = format!("\nPage {}:", page_count);
+                    }
+                }
             }
-            _ => b"[[imposs]]".to_vec(),
-        }),
-    }*/
+        }
+    };
+    Box::pin(StreamReader::new(output_stream))
```

Comment:
I'm not sure whether the use of `Box::pin` is strictly necessary here. It was just the first thing I tried that satisfied the compiler, so I haven't dug further into understanding pinning.

Reply:
that's ok i think, i don't have a deep understanding of pinning / unpinning either